Uv bio enrollment, continued

kaczmarczyck opened the pull request:
Rebases and continues #718 .

The second commits adds the following missing features:

- PIN authentication of BioEnrollment subcommands that require it.
- A compile flag for all new features.
- Cleanup and generalization of the fingerprint API. (Example: `setloglevel` can be part of the `TockEnv` implementation, but shouldn't be included in the API.)
- A more useful fake implementation of the fingerprint API.
- Customization of Veridian's hardcoded numbers.
- Rollback of customizations that are specific to Veridian.
- Whether or not state is held is now left to the implementer of the fingerprint API. BioEnrollment is explicitly not a stateful command according to the specification.
- Don't ignore errors from hardware interaction, but forward them.
- Correct handling of UV retries: Reset everywhere necesary, and don't decrease the retry counter when asking for UV, but only when the user actually presented their finger.
- Removed magic numbers.
- Style matches existing code.
- CBOR parsing is more readable with new structs.
- Unit tests.
- Fixed all comments from the former PR.

Still missing:

- A more complete fake for testing.
- Tests for all subcommands.
- A working TockEnv implementation.

As is, the implementation is not usable on hardware.

Moves a lot of the code from the first commit around, so looking at these commits in isolation is harder than reviewing their sum.

bradjc committed:
Add fingerprint and UV support

kaczmarczyck committed:
Refactors fingerprint code

Introduces a compile flag to toggle fingerprint code.
Makes data structures and style match the library's current standard.
Removes some undesired customization and debug features.

ia0 commented on libraries/opensk/src/api/persist.rs:
```diff
@@ -386,6 +418,102 @@ pub trait Persist {
+            };
+            let mut data = Vec::new();
+            cbor_write(cbor::Value::from(template_info), &mut data)?;
+            self.insert(key, &mut data)?;
```
The `mut` seems weird.
```suggestion
            self.insert(key, &data)?;
```

kaczmarczyck replied:
I was surprised that this wasn't caught by clippy, and indeed, the new feature wasn't covered yet. Fixed all other clippy findings too.

ia0 commented on libraries/opensk/src/api/persist.rs:
```diff
@@ -386,6 +418,102 @@ pub trait Persist {
+                let template_info = TemplateInfo::try_from(cbor_value)?;
+                if &template_info.template_id == template_id {
+                    self.remove(key)?;
+                    return Ok(());
```
This only removes the first such template, but there could be more. In `store_template_id` we don't seem to check for duplicates.

kaczmarczyck replied:
This is supposedly an invariant of the fingerprint module. I made it explicit now, documented it and return an internal error in case of duplicates.
So remove is untouched, and `store_` is adapted.

ia0 commented on libraries/opensk/src/api/persist.rs:
```diff
@@ -386,6 +418,102 @@ pub trait Persist {
+                    .map_err(|_| Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR)?;
+                let template_info = TemplateInfo::try_from(cbor_value)?;
+                if &template_info.template_id == template_id {
+                    return Ok(template_info.template_friendly_name);
```
Same here which suggests there should be an invariant that there's at most one template per template id.

kaczmarczyck replied:
Won't fix because of above introduced invariant.

ia0 commented on libraries/opensk/src/api/persist.rs:
```diff
@@ -386,6 +418,102 @@ pub trait Persist {
+                    template_info.template_friendly_name = Some(String::from(friendly_name));
+                    let mut data = Vec::new();
+                    cbor_write(cbor::Value::from(template_info), &mut data)?;
+                    self.insert(key, &mut data)?;
```
```suggestion
                    self.insert(key, &data)?;
```

kaczmarczyck replied:
Done, see above wrt clippy

ia0 commented on libraries/opensk/src/api/persist.rs:
```diff
@@ -386,6 +418,102 @@ pub trait Persist {
+    ///
+    /// Returns `Err(CTAP2_ERR_INVALID_OPTION)` if the template was not found.
+    #[cfg(feature = "fingerprint")]
+    fn store_friendly_name(&mut self, template_id: &[u8], friendly_name: &str) -> CtapResult<()> {
```
We could take a `String` as argument since we unconditionally convert it to a `String` on successful operation.

kaczmarczyck replied:
Done

ia0 commented on libraries/opensk/src/api/persist.rs:
```diff
@@ -386,6 +418,102 @@ pub trait Persist {
+    ///
+    /// Returns `Err(CTAP2_ERR_FP_DATABASE_FULL)` if the storage is full.
+    #[cfg(feature = "fingerprint")]
+    fn store_template_id(&mut self, template_id: &[u8]) -> CtapResult<()> {
```
We could take `Vec<u8>` directly since we unconditionally convert it on success.

kaczmarczyck replied:
Done

ia0 commented on libraries/opensk/src/api/persist/keys.rs:
```diff
@@ -103,6 +103,16 @@ make_partition! {
 
+    /// Stored friendly names for enrolled fingerprints.
+    #[cfg(feature = "fingerprint")]
+    FRIENDLY_NAMES = 2031..2036;
```
Let's not leave an empty key if we don't need to. If we want to reserve it, we can `_RESERVED_FRIENDLY_NAME = 2036` like for credentials.
```suggestion
    FRIENDLY_NAMES = 2031..2037;
```

kaczmarczyck replied:
Makes sense, done.

ia0 reviewed (commented):
I only reviewed:
- libraries/opensk/src/api/persist.rs
- libraries/opensk/src/api/persist/key.rs
- libraries/opensk/src/ctap/storage.rs

kaczmarczyck committed:
Makes template IDs unique

Also applies clippy lints.

ia0 commented on run_desktop_tests.sh:
```diff
@@ -54,7 +54,7 @@ echo "Running Clippy lints..."
 cargo clippy --lib --tests --bins --benches --features std,"$MOST_FEATURES" -- -D warnings
 (cd libraries/opensk && cargo clippy --features std -- -D warnings)
-(cd libraries/opensk && cargo clippy --features std,config_command,debug_ctap,with_ctap1,vendor_hid,ed25519 -- -D warnings)
+(cd libraries/opensk && cargo clippy --features std,config_command,debug_ctap,with_ctap1,vendor_hid,ed25519,fingerprint -- -D warnings)
```
What about `std,"$MOST_FEATURES"` like 2 lines above?

kaczmarczyck replied:
That doesn't work since these are features from the base Cargo.toml, and here, clippy is running inside `libraries/opensk`. But I could use `--all-features`, and the only difference would be `fuzz`. I don't remember why I explicitly omitted `fuzz` when I wrote this. But it works either way.

ia0 replied:
I see, sounds good.

ia0 reviewed (dismissed):
LGTM for the storage-related part.

kaczmarczyck committed:
Use all features for clippy, to reduce maintenance

bradjc reviewed (approved):
I just reviewed the changes and they all look good to me.

kaczmarczyck commented:
/gemini review

kaczmarczyck commented on libraries/opensk/src/ctap/fingerprint.rs:
```diff
@@ -0,0 +1,407 @@
+        match response.unwrap() {
+            ResponseData::AuthenticatorBioEnrollment(Some(response)) => {
+                assert!(response.modality.is_none());
+                assert!(response.fingerprint_kind.is_none());
+                assert!(response.max_capture_samples_required_for_enroll.is_none());
```
The `internal_retry` variable is used in `libraries/opensk/src/ctap/client_pin.rs:394`.
The marked lines of this comment seem unrelated.

kaczmarczyck commented on libraries/opensk/src/api/customization.rs:
```diff
@@ -446,6 +524,31 @@ pub fn is_valid(customization: &impl Customization) -> bool {
+        }
+
+        // Maximum internal UV attemps must be in range [1, `max_uv_retries`].
+        if customization.max_uv_attempts_for_internal_retries() < 1
+            || customization.max_uv_attempts_for_internal_retries() > customization.max_uv_retries()
+        {
```
That seems to be exactly what the check does, and setting different default values makes the test pass or fail according to expectation.

kaczmarczyck commented on libraries/opensk/src/api/customization.rs:
```diff
@@ -446,6 +524,31 @@
+        {
+            return false;
+        }
+        if customization.preferred_platform_uv_attempts() != 1
+            && customization.max_uv_attempts_for_internal_retries() > 5
+        {
```
Again, seems to be working as intended. I added a comment above the check

In unrelated news, when I changed the default values, I noticed that a different test has a bug for some non-default values, so I split it up and fixed it. The problem was that the UV attempts were not reset between calls.

kaczmarczyck committed:
Fixes misfiring unit test

For some non-default customization settings, the test would give a false
negative.
